File: frontend/src/lib/api/metricsDashboard.ts
import { publicApiBaseUrl } from '$lib/api/config';

async function readJson<T>(res: Response): Promise<T> {
	if (!res.ok) {
		const text = await res.text();
		throw new Error(text || `HTTP ${res.status}`);
	}
	return (await res.json()) as T;
}

export type MetricsSummary = {
	total_cost_usd: string;
	total_tokens: number;
	total_runs: number;
	repo_count: number;
};

export type RepoMetricsRow = {
	repo_id: number;
	repo_name: string;
	total_cost_usd: string;
	total_tokens: number;
	run_count: number;
};

export type AgentMetricsRow = {
	agent: string;
	total_cost_usd: string;
	total_tokens: number;
	run_count: number;
};

export type AgentRunRow = {
	id: number;
	repo_id: number;
	repo_name: string;
	commit_sha: string;
	agent: string;
	model: string;
	input_tokens: number;
	output_tokens: number;
	total_tokens: number;
	cost_usd: string;
	ran_at: string;
};

export type MetricsDashboardResponse = {
	summary: MetricsSummary;
	by_repo: RepoMetricsRow[];
	by_agent: AgentMetricsRow[];
	recent_runs: AgentRunRow[];
};

export async function fetchMetricsDashboard(repoId?: number): Promise<MetricsDashboardResponse> {
	const params = new URLSearchParams();
	if (repoId != null) {
		params.set('repo_id', String(repoId));
	}
	const query = params.toString();
	const url = `${publicApiBaseUrl()}/metrics${query ? `?${query}` : ''}`;
	return readJson<MetricsDashboardResponse>(await fetch(url));
}

File: api/src/specvsreality_api/schemas/metrics_dashboard.py
"""Schemas for the global LLM metrics dashboard."""

from __future__ import annotations

from datetime import datetime
from decimal import Decimal

from pydantic import BaseModel, ConfigDict, Field


class MetricsSummary(BaseModel):
    model_config = ConfigDict(extra="forbid")

    total_cost_usd: Decimal
    total_tokens: int
    total_runs: int
    repo_count: int


class RepoMetricsRow(BaseModel):
    model_config = ConfigDict(extra="forbid")

    repo_id: int
    repo_name: str
    total_cost_usd: Decimal
    total_tokens: int
    run_count: int


class AgentMetricsRow(BaseModel):
    model_config = ConfigDict(extra="forbid")

    agent: str
    total_cost_usd: Decimal
    total_tokens: int
    run_count: int


class AgentRunRow(BaseModel):
    model_config = ConfigDict(extra="forbid")

    id: int
    repo_id: int
    repo_name: str
    commit_sha: str
    agent: str
    model: str
    input_tokens: int
    output_tokens: int
    total_tokens: int
    cost_usd: Decimal
    ran_at: datetime


class MetricsDashboardResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")

    summary: MetricsSummary
    by_repo: list[RepoMetricsRow]
    by_agent: list[AgentMetricsRow]
    recent_runs: list[AgentRunRow] = Field(default_factory=list)
